src/serial.rs

```diff
@@ -1398,388 +1399,449 @@ pub struct Tx<USART> {
+        };
+        self
+    }
+}
 
-            impl blocking::serial::Write<u8> for Tx<$USARTX> {
-                type Error = Error;
-
-                fn bwrite_all(&mut self, bytes: &[u8]) -> Result<(), Self::Error> {
-                    for &b in bytes {
-                        loop {
-                            match self.write(b) {
-                                Err(nb::Error::WouldBlock) => continue,
-                                Err(nb::Error::Other(err)) => return Err(err),
-                                Ok(()) => break,
-                            }
-                        }
-                    }
-                    Ok(())
+mod private {
+    pub trait Sealed {}
+}
+
+// Implemented by all USART instances
+pub trait Instance: private::Sealed {
+    #[doc(hidden)]
+    fn ptr() -> *const stm32f4::stm32f405::uart4::RegisterBlock;
+    #[doc(hidden)]
+    fn pclk_freq(clocks: &Clocks) -> u32;
+    #[doc(hidden)]
+    fn enable_clock(rcc: &crate::stm32::rcc::RegisterBlock);
+    #[doc(hidden)]
+    fn stopbits(bits: config::StopBits) -> stm32f4::stm32f405::uart4::cr2::STOP_A;
+}
+
+macro_rules! halUsart {
+    ($(
+        $USARTX:ident: ($usartX:ident, $apbXenr:ident, $rcc_bit:expr, $usartXen:ident, $pclkX:ident),
+    )+) => {
+        $(
+            impl private::Sealed for $USARTX {}
+            impl Instance for $USARTX {
+                fn ptr() -> *const stm32f4::stm32f405::uart4::RegisterBlock {
+                    unsafe { mem::transmute($USARTX::ptr()) }
                 }
 
-                fn bflush(&mut self) -> Result<(), Self::Error> {
-                    loop {
-                        match self.flush() {
-                            Ok(()) => return Ok(()),
-                            Err(nb::Error::WouldBlock) => continue,
-                            Err(nb::Error::Other(err)) => return Err(err),
-                        }
-                    }
+                fn pclk_freq(clocks: &Clocks) -> u32 {
+                    clocks.$pclkX().0
                 }
-            }
 
-            impl<PINS> blocking::serial::Write<u8> for Serial<$USARTX, PINS> {
-                type Error = Error;
+                fn enable_clock(rcc: &crate::stm32::rcc::RegisterBlock) {
+                    use crate::bb;
 
-                fn bwrite_all(&mut self, bytes: &[u8]) -> Result<(), Self::Error> {
-                    let mut tx: Tx<$USARTX> = Tx {
-                        _usart: PhantomData,
-                    };
-                    tx.bwrite_all(bytes)
+                    unsafe { bb::set(&rcc.$apbXenr, $rcc_bit) };
                 }
 
-                fn bflush(&mut self) -> Result<(), Self::Error> {
-                    let mut tx: Tx<$USARTX> = Tx {
-                        _usart: PhantomData,
-                    };
-                    tx.bflush()
+                fn stopbits(bits: config::StopBits) -> stm32f4::stm32f405::uart4::cr2::STOP_A {
```

Comment:
Maybe you can get rid of the transmute here if you use `crate::pac` instead of `stm32f4::stm32f405` ?

Reply:
No, because the types are different.